moveit_core/robot_state/src/robot_state.cpp

```diff
@@ -1357,91 +1357,78 @@ bool RobotState::getJacobian(const JointModelGroup* group, const LinkModel* link
-        jacobian.block<3, 1>(0, joint_index) =
-            jacobian.block<3, 1>(0, joint_index) + joint_axis.cross(point_transform - joint_transform.translation());
-        jacobian.block<3, 1>(3, joint_index) = jacobian.block<3, 1>(3, joint_index) + joint_axis;
-      }
-      else if (pjm->getType() == moveit::core::JointModel::PRISMATIC)
-      {
-        joint_axis = joint_transform.linear() * static_cast<const moveit::core::PrismaticJointModel*>(pjm)->getAxis();
-        jacobian.block<3, 1>(0, joint_index) = jacobian.block<3, 1>(0, joint_index) + joint_axis;
-      }
-      else if (pjm->getType() == moveit::core::JointModel::PLANAR)
-      {
-        joint_axis = joint_transform * Eigen::Vector3d(1.0, 0.0, 0.0);
-        jacobian.block<3, 1>(0, joint_index) = jacobian.block<3, 1>(0, joint_index) + joint_axis;
-        joint_axis = joint_transform * Eigen::Vector3d(0.0, 1.0, 0.0);
-        jacobian.block<3, 1>(0, joint_index + 1) = jacobian.block<3, 1>(0, joint_index + 1) + joint_axis;
-        joint_axis = joint_transform * Eigen::Vector3d(0.0, 0.0, 1.0);
-        jacobian.block<3, 1>(0, joint_index + 2) = jacobian.block<3, 1>(0, joint_index + 2) +
-                                                   joint_axis.cross(point_transform - joint_transform.translation());
-        jacobian.block<3, 1>(3, joint_index + 2) = jacobian.block<3, 1>(3, joint_index + 2) + joint_axis;
-      }
-      else
-        RCLCPP_ERROR(LOGGER, "Unknown type of joint in Jacobian computation");
+      Eigen::Vector3d axis_wrt_origin =
+          origin_pose_link.linear() * static_cast<const moveit::core::RevoluteJointModel*>(joint_model)->getAxis();
+      jacobian.block<3, 1>(0, i) = axis_wrt_origin.cross(tip_point - origin_pose_link.translation());
+      jacobian.block<3, 1>(3, i) = axis_wrt_origin;
+    }
+    else if (joint_model->getType() == moveit::core::JointModel::PRISMATIC)
+    {
+      Eigen::Vector3d axis_wrt_origin =
+          origin_pose_link.linear() * static_cast<const moveit::core::PrismaticJointModel*>(joint_model)->getAxis();
+      jacobian.block<3, 1>(0, i) = axis_wrt_origin;
+      jacobian.block<3, 1>(3, i) = Eigen::Vector3d::Zero();
     }
-    if (pjm == root_joint_model)
-      break;
-    link = pjm->getParentLinkModel();
+    else if (joint_model->getType() == moveit::core::JointModel::PLANAR)
+    {
+      jacobian.block<3, 1>(0, i) = origin_pose_link.linear() * Eigen::Vector3d(1.0, 0.0, 0.0);
+      jacobian.block<3, 1>(0, i + 1) = origin_pose_link.linear() * Eigen::Vector3d(0.0, 1.0, 0.0);
+      jacobian.block<3, 1>(0, i + 2) =
+          (origin_pose_link.linear() * Eigen::Vector3d(0.0, 0.0, 1.0)).cross(tip_point - origin_pose_link.translation());
+      jacobian.block<3, 1>(3, i + 2) = origin_pose_link.linear() * Eigen::Vector3d(0.0, 0.0, 1.0);
+    }
+    else
+    {
+      RCLCPP_ERROR(LOGGER, "Unknown type of joint in Jacobian computation");
```

Comment:
Should we return `false` in here instead of proceeding?

Reply:
Yes, I think so. Changed.